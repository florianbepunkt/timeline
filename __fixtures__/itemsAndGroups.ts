import type { TimelineGroupBase, TimelineItemBase } from "../src";

export const items: TimelineItemBase[] = [
  {
    id: "0",
    group: "1",
    startTime: 1540540000000,
    endTime: 1540546803877,
    canMove: true,
    canResize: "both",
  },
  {
    id: "5",
    group: "1",
    startTime: 1540532800000,
    endTime: 1540554003877,
    canMove: true,
    canResize: "both",
    className: "",
  },
  {
    id: "6",
    group: "1",
    startTime: 1540550800000,
    endTime: 1540575603877,
    canMove: true,
    canResize: "both",
    className: "",
  },
  {
    id: "1",
    group: "1",
    startTime: 1540570000000,
    endTime: 1540584875919,
    canMove: true,
    canResize: "both",
  },
  {
    id: "2",
    group: "1",
    startTime: 1540620000000,
    endTime: 1540640397548,
    canMove: true,
    canResize: "both",
    className: "",
  },
  {
    id: "3",
    group: "3",
    startTime: 1540656000000,
    endTime: 1540676397548,
    canMove: true,
    canResize: "both",
    className: "",
  },
];

export const groups: TimelineGroupBase[] = [
  { id: "1", title: "1" },
  { id: "2", title: "2" },
  { id: "3", title: "3" },
];
