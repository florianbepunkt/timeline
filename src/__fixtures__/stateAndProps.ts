import { items } from "./itemsAndGroups";

export const visibleTimeStart = 1540501200000;
export const visibleTimeEnd = 1540587600000;

export const props = {
  lineHeight: 30,
  stackItems: true,
  itemHeightRatio: 0.75,
  visibleTimeEnd,
  visibleTimeStart,
};

export const propsNoStack = {
  ...props,
  stackItems: false,
};

export const state = {
  draggingItem: undefined,
  dragTime: null,
  resizingItem: null,
  resizingEdge: null,
  resizeTime: null,
  newGroupOrder: null,
  canvasTimeStart: 1540414800000,
  visibleTimeEnd: visibleTimeEnd,
  visibleTimeStart: visibleTimeStart,
  canvasTimeEnd: 1540674000000,
  width: 1000,
};

//offset 1 hour
const timeOffset = 1 * 60 * 60 * 1000;

export const stateMoveItem = {
  ...state,
  draggingItem: items[0].id,
  dragTime: items[0].startTime + timeOffset,
  newGroupOrder: 0,
};
export const stateResizeItemLeft = {
  ...state,
  resizingItem: items[0].id,
  resizingEdge: "left",
  resizeTime: items[0].startTime + timeOffset,
  newGroupOrder: 0,
};

export const stateResizeItemRight = {
  ...state,
  resizingItem: items[0].id,
  resizingEdge: "right",
  resizeTime: items[0].endTime + timeOffset,
  newGroupOrder: 0,
};
