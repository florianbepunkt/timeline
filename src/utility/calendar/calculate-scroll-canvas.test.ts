import { calculateScrollCanvas } from "./calendar.js";
import { describe, expect, test } from "vitest";
import { items, groups } from "../../__fixtures__/itemsAndGroups.js";
import { props, state, visibleTimeStart, visibleTimeEnd } from "../../__fixtures__/stateAndProps.js";

describe("calculateScrollCanvas", () => {
  test("should calculate new scroll state", () => {
    const newStartTime = visibleTimeStart + 13 * 60 * 60 * 1000;
    const newEndTime = visibleTimeEnd + visibleTimeStart + 13 * 60 * 60 * 1000;
    const result = calculateScrollCanvas(newStartTime, newEndTime, false, items, groups, props, state);
    expect(result).toHaveProperty("visibleTimeStart");
    expect(result).toHaveProperty("visibleTimeEnd");
    expect(result).toHaveProperty("dimensionItems");
  });

  test("should calculate new scroll state correctly", () => {
    const newStartTime = visibleTimeStart + 13 * 60 * 60 * 1000;
    const newEndTime = visibleTimeEnd + 13 * 60 * 60 * 1000;
    const result = calculateScrollCanvas(newStartTime, newEndTime, false, items, groups, props, state);
    expect(result).toMatchSnapshot();
  });

  test("should skip new calculation if new visible start and visible end in canvas", () => {
    const newStartTime = visibleTimeStart + 1 * 60 * 60 * 1000;
    const newEndTime = visibleTimeEnd + 1 * 60 * 60 * 1000;
    const result = calculateScrollCanvas(newStartTime, newEndTime, false, items, groups, props, state);
    expect(result).toHaveProperty("visibleTimeStart");
    expect(result).toHaveProperty("visibleTimeEnd");
    expect(result).not.toHaveProperty("dimensionItems");
  });

  test("should force new calculation", () => {
    const newStartTime = visibleTimeStart + 1 * 60 * 60 * 1000;
    const newEndTime = visibleTimeEnd + 1 * 60 * 60 * 1000;
    const result = calculateScrollCanvas(
      newStartTime.valueOf(),
      newEndTime.valueOf(),
      true,
      items,
      groups,
      props,
      state
    );
    expect(result).toHaveProperty("visibleTimeStart");
    expect(result).toHaveProperty("visibleTimeEnd");
    expect(result).toHaveProperty("dimensionItems");
  });

  test("should calculate new state if zoom changed ", () => {
    const newStartTime = visibleTimeStart;
    const newEndTime = visibleTimeEnd + 1 * 60 * 60 * 1000;
    const result = calculateScrollCanvas(newStartTime, newEndTime, false, items, groups, props, state);
    expect(result).toHaveProperty("visibleTimeStart");
    expect(result).toHaveProperty("visibleTimeEnd");
    expect(result).toHaveProperty("dimensionItems");
  });

  test("should calculate new state if zoom changed correctly", () => {
    const newStartTime = visibleTimeStart;
    const newEndTime = visibleTimeEnd + 1 * 60 * 60 * 1000;
    const result = calculateScrollCanvas(newStartTime, newEndTime, false, items, groups, props, state);
    expect(result).toMatchSnapshot();
  });
});
