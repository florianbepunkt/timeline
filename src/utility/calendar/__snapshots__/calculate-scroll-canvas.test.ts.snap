// Vitest Snapshot v1

exports[`calculateScrollCanvas > should calculate new scroll state correctly 1`] = `
{
  "canvasTimeEnd": 1540720800000,
  "canvasTimeStart": 1540461600000,
  "dimensionItems": [
    {
      "dimensions": {
        "collisionLeft": 1540540000000,
        "collisionWidth": 6803877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": undefined,
        "left": 907.4074074074074,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": NaN,
        "width": 78.74857638888886,
      },
      "id": "0",
    },
    {
      "dimensions": {
        "collisionLeft": 1540532800000,
        "collisionWidth": 21203877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": undefined,
        "left": 824.074074074074,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": NaN,
        "width": 245.4152430555556,
      },
      "id": "5",
    },
    {
      "dimensions": {
        "collisionLeft": 1540550800000,
        "collisionWidth": 24803877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": undefined,
        "left": 1032.4074074074074,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": NaN,
        "width": 287.08190972222224,
      },
      "id": "6",
    },
    {
      "dimensions": {
        "collisionLeft": 1540570000000,
        "collisionWidth": 14875919,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": undefined,
        "left": 1254.6296296296296,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": NaN,
        "width": 172.1749884259259,
      },
      "id": "1",
    },
    {
      "dimensions": {
        "collisionLeft": 1540620000000,
        "collisionWidth": 20397548,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": undefined,
        "left": 1833.3333333333333,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": NaN,
        "width": 236.08273148148123,
      },
      "id": "2",
    },
    {
      "dimensions": {
        "collisionLeft": 1540656000000,
        "collisionWidth": 20397548,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": undefined,
        "left": 2250,
        "order": {
          "group": {
            "id": "3",
            "title": "3",
          },
          "index": 2,
        },
        "stack": true,
        "top": NaN,
        "width": 236.08273148148146,
      },
      "id": "3",
    },
  ],
  "groupHeights": [
    30,
    30,
    30,
  ],
  "groupTops": [
    0,
    30,
    60,
  ],
  "height": 90,
  "visibleTimeEnd": 1540634400000,
  "visibleTimeStart": 1540548000000,
}
`;

exports[`calculateScrollCanvas > should calculate new state if zoom changed correctly 1`] = `
{
  "canvasTimeEnd": 1540681200000,
  "canvasTimeStart": 1540411200000,
  "dimensionItems": [
    {
      "dimensions": {
        "collisionLeft": 1540540000000,
        "collisionWidth": 6803877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": undefined,
        "left": 1431.111111111111,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": NaN,
        "width": 75.59863333333351,
      },
      "id": "0",
    },
    {
      "dimensions": {
        "collisionLeft": 1540532800000,
        "collisionWidth": 21203877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": undefined,
        "left": 1351.111111111111,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": NaN,
        "width": 235.5986333333335,
      },
      "id": "5",
    },
    {
      "dimensions": {
        "collisionLeft": 1540550800000,
        "collisionWidth": 24803877,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": undefined,
        "left": 1551.111111111111,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": NaN,
        "width": 275.5986333333335,
      },
      "id": "6",
    },
    {
      "dimensions": {
        "collisionLeft": 1540570000000,
        "collisionWidth": 14875919,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": undefined,
        "left": 1764.4444444444446,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": NaN,
        "width": 165.28798888888878,
      },
      "id": "1",
    },
    {
      "dimensions": {
        "collisionLeft": 1540620000000,
        "collisionWidth": 20397548,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": undefined,
        "left": 2320,
        "order": {
          "group": {
            "id": "1",
            "title": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": NaN,
        "width": 226.6394222222225,
      },
      "id": "2",
    },
    {
      "dimensions": {
        "collisionLeft": 1540656000000,
        "collisionWidth": 20397548,
        "extraSpaceLeft": null,
        "extraSpaceRight": null,
        "height": undefined,
        "left": 2720,
        "order": {
          "group": {
            "id": "3",
            "title": "3",
          },
          "index": 2,
        },
        "stack": true,
        "top": NaN,
        "width": 226.6394222222225,
      },
      "id": "3",
    },
  ],
  "groupHeights": [
    30,
    30,
    30,
  ],
  "groupTops": [
    0,
    30,
    60,
  ],
  "height": 90,
  "visibleTimeEnd": 1540591200000,
  "visibleTimeStart": 1540501200000,
}
`;
